docs/recipes/MigratingToI18Next.md

```diff
@@ -0,0 +1,164 @@
+  initI18n().then(() => setIsI18nInitialized(true));
+}, []);
+```
+
+Additionally, consider including the new state variable in the rendering condition for the app.
+
+```js
+// error-line-start
+if (!rehydrated || !isNavigationStateRestored || (!areFontsLoaded && !fontLoadError)) {
+// error-line-end
+// success-line-start
+if (!rehydrated || !isNavigationStateRestored || !isI18nInitialized || (!areFontsLoaded && !fontLoadError)) {
+// success-line-end
+  return null
+}
+```
+
+This ensures that your app will wait until `react-i18next` is fully initialized before continuing, preventing any issues with missing translations.
+
+## Step 3: Update the i18n initialization method
+
+Next, update your i18n initialization to use `react-i18next`, which also includes RTL (right-to-left) language support and handles locale selection. In a Ingnite generated project, this is located in `app/i18n/i18n.ts`.
+
+```js
+
+import * as i18next from "i18next"
+
+const initI18n = async () => {
+  await i18n.use(initReactI18next).init({
+    resources,
+    lng: pickSupportedLocale()?.languageTag || fallbackLocale,
+    fallbackLng: fallbackLocale,
+    interpolation: { escapeValue: false },
+  });
+
+  const locale = pickSupportedLocale();
+  if (locale?.textDirection === 'rtl') {
+    I18nManager.allowRTL(true);
+    isRTL = true;
+  } else {
+    I18nManager.allowRTL(false);
+    isRTL = false;
+  }
+
+  return i18n;
+};
+```
+
+This ensures that supported locales are chosen based on the device’s settings, and RTL is correctly applied when necessary. For more on detail on these changes, check the [this PR](https://github.com/infinitered/ignite/pull/2770).
+
+## Step 4: Add intl-pluralrules for react-i18next and JSON v4
+
+To support pluralization and `react-i18next`'s JSON v4 format, you’ll need to add the `intl-pluralrules` package:
+
+```bash
+yarn add intl-pluralrules
+```
+
+Make sure to import this package into your i18n configuration file (`app/i18n/i18n.ts`):
+
+```js
+import 'intl-pluralrules';
+```
+
+## Step 5: Update the translate function
+
+The next step is to replace your old translate function with the one provided by `react-i18next`. This is located in `app/i18n/translate.ts`:
+
+```js
+// error-line
+import { TranslateOptions } from "i18n-js"
+// success-line
+import { TOptions } from "i18next"
+
+// error-line-start
+export function translate(key: TxKeyPath, options?: TranslateOptions): string {
+  return i18n.t(key, options)
+// error-line-end
+// success-line-start
+export function translate(key, options) {
+  return i18n.isInitialized ? i18n.t(key, options) : key;
+// success-line-end
+}
+```
+
+## Step 6: Update translation keys from dots (.) to colons (:)
+
+`react-i18next` uses different types of separators for translation keys. Colons (:) are used for first-level translations within an object, while dots (.) are used for nested translations. As a result, you’ll need to update all translation keys in your app accordingly. For example:
```

Comment:
In doing this dot notation to colon notation, you will receive errors on any `tx` prop until you update the types in i18n.ts to

```tsx
type RecursiveKeyOf<TObj extends object> = {
  [TKey in keyof TObj & (string | number)]: RecursiveKeyOfHandleValue<TObj[TKey], `${TKey}`, true>
}[keyof TObj & (string | number)]

type RecursiveKeyOfInner<TObj extends object> = {
  [TKey in keyof TObj & (string | number)]: RecursiveKeyOfHandleValue<TObj[TKey], `${TKey}`, false>
}[keyof TObj & (string | number)]

type RecursiveKeyOfHandleValue<
  TValue,
  Text extends string,
  IsFirstLevel extends boolean,
> = TValue extends any[]
  ? Text
  : TValue extends object
    ? IsFirstLevel extends true
      ? Text | `${Text}:${RecursiveKeyOfInner<TValue>}`
      : Text | `${Text}.${RecursiveKeyOfInner<TValue>}`
    : Text
 ```

Reply:
So after showing this change, I'd probably show an example of changing some props as you did with the translate function below (but we generally don't use that one as much in the boilerplate). Maybe an example from the `<WelcomeScreen />` since it's pretty straight forward, for example:

```tsx
  useHeader(
    {
      rightTx: "common:logOut",
      onRightPress: logout,
    },
    [logout],
  )
  ```
  
  ```tsx
  return (
    <View style={$container}>
      <View style={$topContainer}>
        <Image style={$welcomeLogo} source={welcomeLogo} resizeMode="contain" />
        <Text
          testID="welcome-heading"
          style={$welcomeHeading}
          tx="welcomeScreen:readyForLaunch"
          preset="heading"
        />
        <Text tx="welcomeScreen:exciting" preset="subheading" />
        <Image style={$welcomeFace} source={welcomeFace} resizeMode="contain" />
      </View>

      <View style={[$bottomContainer, $bottomContainerInsets]}>
        <Text tx="welcomeScreen:postscript" size="md" />

        <Button
          testID="next-screen-button"
          preset="reversed"
          tx="welcomeScreen:letsGo"
          onPress={goNext}
        />
      </View>
    </View>
  )
})
```

Reply:
> you will receive errors on any tx prop until you update the types in i18n.ts to

I never looked into this, but I was wondering why we kept types that were defined when using the previous library. I would've expected we'd be able to use types defined by the new library.